### DISCUSSION TITLE
Performance Tips - 500M vectors HNSW Index

### DISCUSSION THREAD

**Discussion by rgoularti** [milvus-io/milvus#19189] [Q&A and General discussion] [answered]
Hello Milvus community!

I’m doing some benchmarks on Milvus Cluster v2.1.1 installed with Helm and deployed on GKE, and got a very slow search response time and QPS.

My use case:

- Single collection with 500M random vectors / 512 dim.
- HNSW Index
    - Metric Type: L2
    - Params: M=24, efConstruction=256
- Data loaded on 45 Query Nodes with 8vCPUs 32Gb ram each
- Default milvus configurations ([https://milvus.io/docs/v2.1.x/configure-helm.md#Download-a-configuration-file](https://milvus.io/docs/v2.1.x/configure-helm.md#Download-a-configuration-file))
- Searching 1 vector against the entire collection returning the top 10 most similar vectors.

With this setup I manage to get around 5 QPS with the average RT of 1700ms, tested with 10 concurrent users using locust.

My goal is to reach a RT below 1s and the higher QPS possible.  

Any performance tips will be highly appreciated.

**Comment by yhmo** [comment]
Did you specify the consistency level when you call the search()?

In our [documentation](https://milvus.io/docs/v2.1.x/search.md), the example uses a "Strong" consistency level:
```
results = collection.search(
	data=[[0.1, 0.2]], 
	anns_field="book_intro", 
	param=search_params, 
	limit=10, 
	expr=None,
	consistency_level="Strong"
)
```

This kind of consistency level has 200ms latency for each search request. You can use "Bounded" or "Eventually" level and try again:
```
results = collection.search(
	data=[[0.1, 0.2]], 
	anns_field="book_intro", 
	param=search_params, 
	limit=10, 
	expr=None,
	consistency_level="Bounded"
)
```

**Reply by rgoularti** [reply]
I'm already using "Bounded" consistency level, but I didn't know that this setting could impact on the latency.
I will be aware of that, thanks for sharing!

**Comment by yhmo** [comment, marked as answer]
Recently we create a new sizing tool to calculate the required memory size for milvus.
https://milvus.io/tools/sizing/

I input your parameters:
![Screenshot from 2022-09-14 10-40-16](https://user-images.githubusercontent.com/2282099/190046568-d9434fcc-0509-4f36-a45f-3181af860d95.png)

It provides recommendations for milvus cluster deployment, you can have a try.

**Reply by rgoularti** [reply]
Great @yhmo !
Based on the sizing tool recommendations I can do some adjustments on my configuration.

I'll comment back  as soon as I run the tests.
Thanks!

**Comment by yhmo** [comment]
Some other reasons for low performance:
1. Not all the data has been indexed.
Use `utility.index_building_progress(collection_name="xxx")` to check whether the indexed row count is equal to total row count.

2. The segments were not well balanced across query nodes.
Observe the CPU usage during concurrent search. If some nodes are busy and some others are idle then the data is not well balanced.

**Comment by yhmo** [comment]
@rgoularti 
Hi, how about the search performance now?
If you still get a low QPS, please let us know. And for further diagnose, the [Milvus monitoring services](https://milvus.io/docs/v2.1.x/monitor.md) is helpful.

**Reply by rgoularti** [reply]
Hi @yhmo !
Sorry for the late response.

Yes, I was able to improve the search performance by following the sizing tool recommendations and, additionally, increasing the `dataCoordinator.segment.maxSize` from 512 to 2048.

I manage to reach 25 QPS under 1000 ms, which is good for my use case.

Also, I tested the In-Memory replicas feature, on smaller datasets, and confirmed a great improvement on the search throughput, being a good option if we need more search capacity. 

Thank you very much!
